{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "independent-europe",
   "metadata": {},
   "source": [
    "#### Summary:\n",
    "In this notebook I will streamline the LAI HMM implementation to be easy to use in various test cases. *** NOTE: *** This notebook is currently designed to be implemented in the datahub team3 directory, not the github repo! This could certainly be streamlined even more, and will be when we test the effect of a range of recombination rates."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "chronic-anderson",
   "metadata": {},
   "outputs": [],
   "source": [
    "import random as rn\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import sys\n",
    "import math"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "unexpected-parts",
   "metadata": {},
   "outputs": [],
   "source": [
    "import datetime"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "corrected-financing",
   "metadata": {},
   "outputs": [],
   "source": [
    "from LAI_hmm_scriptFINAL import HMMOptimalPathLAI, HammingDist, makeSNPseq, standardizeIndices"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "liquid-shannon",
   "metadata": {},
   "outputs": [],
   "source": [
    "from LAI_hmm_script_progress import HMMOptimalPathLAI_progress"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "comparative-target",
   "metadata": {},
   "source": [
    "# Read in Data and Create an HMM Object"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "closed-facility",
   "metadata": {},
   "source": [
    "For all simulated data tests:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "veterinary-plumbing",
   "metadata": {},
   "outputs": [],
   "source": [
    "#define names of the two test populations\n",
    "test_PopA = \"0\"\n",
    "test_PopB = \"1\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "everyday-bleeding",
   "metadata": {},
   "outputs": [],
   "source": [
    "#read in emissions data (simulated df)\n",
    "emission_fp = \"Git/cse284_project/Data/simData_N2_P100_seed518.tsv\" #asumes in team3 dir\n",
    "emission_df = pd.read_csv(emission_fp, sep='\\t',header=0)\n",
    "\n",
    "#renaming the columns (only necessary for simulated data)\n",
    "colnames = [\"POS\",\"0_A\",\"0_C\",\"0_G\",\"0_T\",\"1_A\",\"1_C\",\"1_G\",\"1_T\"]\n",
    "emission_df.columns = colnames\n",
    "#emission_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "double-turkish",
   "metadata": {},
   "source": [
    "Individual (test) specific:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "sophisticated-victoria",
   "metadata": {},
   "outputs": [],
   "source": [
    "#read in sample genotype to test\n",
    "#assumes current dir is the team3 dir\n",
    "genotype_fp = \"Git/cse284_project/Data/simGenome_100_0_0.tsv\"\n",
    "genotype_df = pd.read_csv(genotype_fp, sep='\\t',header = 0)\n",
    "#genotype_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "cutting-ethiopia",
   "metadata": {},
   "outputs": [],
   "source": [
    "#cut down both dfs to only be positions shared in both\n",
    "fin_emission_df, fin_genotype_df = standardizeIndices(emission_df,genotype_df,\"POS\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "decimal-psychiatry",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>POS</th>\n",
       "      <th>0_A</th>\n",
       "      <th>0_C</th>\n",
       "      <th>0_G</th>\n",
       "      <th>0_T</th>\n",
       "      <th>1_A</th>\n",
       "      <th>1_C</th>\n",
       "      <th>1_G</th>\n",
       "      <th>1_T</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.3553</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.6447</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.5719</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.4281</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.7580</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.2420</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.4374</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.5626</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>0.1303</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.8697</td>\n",
       "      <td>0.4271</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.5729</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>0.2157</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.7843</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.3534</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.6466</td>\n",
       "      <td>0.0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>0.2444</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.7556</td>\n",
       "      <td>0.3655</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.6345</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   POS     0_A     0_C     0_G     0_T     1_A     1_C     1_G     1_T\n",
       "0    0  0.0000  0.3553  0.0000  0.6447  0.0000  0.5719  0.0000  0.4281\n",
       "1    1  0.0000  0.7580  0.0000  0.2420  0.0000  0.4374  0.0000  0.5626\n",
       "2    2  0.1303  0.0000  0.0000  0.8697  0.4271  0.0000  0.0000  0.5729\n",
       "3    3  0.2157  0.0000  0.7843  0.0000  0.3534  0.0000  0.6466  0.0000\n",
       "4    4  0.2444  0.0000  0.0000  0.7556  0.3655  0.0000  0.0000  0.6345"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fin_emission_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "noble-costume",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>POS</th>\n",
       "      <th>A1</th>\n",
       "      <th>A2</th>\n",
       "      <th>POP1</th>\n",
       "      <th>POP2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   POS A1 A2  POP1  POP2\n",
       "0    0  C  T     0     0\n",
       "1    1  C  C     0     0\n",
       "2    2  T  T     0     0\n",
       "3    3  A  G     0     0\n",
       "4    4  T  T     0     0"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fin_genotype_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "geographic-reading",
   "metadata": {},
   "outputs": [],
   "source": [
    "#create SNPseq\n",
    "a1_idx = list(genotype_df.columns).index('A1')\n",
    "a2_idx = list(genotype_df.columns).index('A2')\n",
    "snps = makeSNPseq(fin_genotype_df,a1_idx,a2_idx)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "criminal-cleaning",
   "metadata": {},
   "source": [
    "HMM Creation and Initialization"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "immediate-watershed",
   "metadata": {},
   "outputs": [],
   "source": [
    "#set recombination rate\n",
    "recomb = 0.1\n",
    "\n",
    "#create the hmm object\n",
    "test_hmm00 = HMMOptimalPathLAI(test_PopA, test_PopB, fin_emission_df, recomb, snps)\n",
    "\n",
    "#initialize the transition matrix\n",
    "test_hmm00.get_transition_matrix()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "willing-administrator",
   "metadata": {},
   "source": [
    "# Reconstruct Haplotypes with the HMM Model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "floating-contamination",
   "metadata": {},
   "outputs": [],
   "source": [
    "#perform the Viterbi algorithm\n",
    "test_hmm00.get_optimal_path()\n",
    "#print(test_hmm00.optimal_matrix)\n",
    "\n",
    "#reconstruct the most probable path through states\n",
    "test_hmm00.reconstruct_path()\n",
    "#print(test_hmm00.final_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "likely-export",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "HMM_Test_Outputs/simGenome_100_0_0_HMMoutput.tsv\n"
     ]
    }
   ],
   "source": [
    "#convert the path to the desired output format (pass the string to use for CHR col)\n",
    "path_df = test_hmm00.output_path(\"21\")\n",
    "#print(path_df.head())\n",
    "\n",
    "#save to a file (also assumes are currently in team3 directory)\n",
    "output_fp = \"HMM_Test_Outputs/\" + genotype_fp.split(\"/\")[3][:-4] + \"_HMMoutput.tsv\"\n",
    "print(output_fp)\n",
    "path_df.to_csv(output_fp,sep='\\t',header=True,index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "cooked-vinyl",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHR</th>\n",
       "      <th>POS</th>\n",
       "      <th>POP1</th>\n",
       "      <th>POP2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>21</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>21</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>21</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>21</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>21</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  CHR  POS POP1 POP2\n",
       "0  21    0    0    0\n",
       "1  21    1    0    0\n",
       "2  21    2    0    0\n",
       "3  21    3    0    0\n",
       "4  21    4    0    0"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "path_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "stunning-amino",
   "metadata": {},
   "outputs": [],
   "source": [
    "#alternative ultra streamlined version (create HMM --> get optimal path)\n",
    "\n",
    "# recomb = 0.1\n",
    "# test_hmm00 = HMMOptimalPathLAI(test_PopA, test_PopB, fin_emission_df, recomb, snps)\n",
    "# test_hmm00.get_transition_matrix()\n",
    "# test_hmm00.get_optimal_path()\n",
    "# test_hmm00.reconstruct_path()\n",
    "# path_df = test_hmm00.output_path(\"21\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "stable-client",
   "metadata": {},
   "source": [
    "# As One Function"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "sonic-insertion",
   "metadata": {},
   "outputs": [],
   "source": [
    "#define names of the two test populations\n",
    "test_PopA = \"0\"\n",
    "test_PopB = \"1\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "confused-southeast",
   "metadata": {},
   "outputs": [],
   "source": [
    "#read in emissions data (simulated df)\n",
    "emission_fp = \"Git/cse284_project/Data/simData_N2_P100_seed518.tsv\" #asumes in team3 dir\n",
    "test_emission_df = pd.read_csv(emission_fp, sep='\\t',header=0)\n",
    "\n",
    "#renaming the columns (only necessary for simulated data)\n",
    "colnames = [\"POS\",\"0_A\",\"0_C\",\"0_G\",\"0_T\",\"1_A\",\"1_C\",\"1_G\",\"1_T\"]\n",
    "test_emission_df.columns = colnames\n",
    "#emission_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "animal-street",
   "metadata": {},
   "outputs": [],
   "source": [
    "def HMM_implementation_wrapper(popA, popB, genotype_fp, emission_df, recomb, output_fp, chrm):\n",
    "    #read in sample genotype to test -- assumes current dir is the team3 dir\n",
    "    genotype_df = pd.read_csv(genotype_fp, sep='\\t',header = 0)\n",
    "\n",
    "    #cut down both dfs to only be positions shared in both\n",
    "    fin_emission_df, fin_genotype_df = standardizeIndices(emission_df, genotype_df, \"POS\")\n",
    "    \n",
    "    #create SNPseq\n",
    "    a1_idx = list(fin_genotype_df.columns).index('A1')\n",
    "    a2_idx = list(fin_genotype_df.columns).index('A2')\n",
    "    snps = makeSNPseq(fin_genotype_df,a1_idx,a2_idx)\n",
    "    \n",
    "    #create the hmm object\n",
    "    hmm = HMMOptimalPathLAI(popA, popB, fin_emission_df, recomb, snps)\n",
    "    hmm.get_transition_matrix()\n",
    "    \n",
    "    #perform the Viterbi algorithm and reconstruct the most probable path through states\n",
    "    hmm.get_optimal_path()\n",
    "    hmm.reconstruct_path()\n",
    "\n",
    "    #convert the path to the desired output format (pass the string to use for CHR col)\n",
    "    path_df = hmm.output_path(chrm)\n",
    "    path_df.to_csv(output_fp,sep='\\t',header=True,index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "julian-calculator",
   "metadata": {},
   "source": [
    "# Running the HMM on Short Simulated Genotypes with Various Recombination Rates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "extraordinary-situation",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "#collect all the file names in one file\n",
    "#touch /home/hmummey/teams/CSE284_SP21_A00/team3/HMM_Test_Outputs/simulated_files.txt\n",
    "cd /home/hmummey/teams/CSE284_SP21_A00/team3/Git/cse284_project/Data\n",
    "\n",
    "echo /home/hmummey/teams/CSE284_SP21_A00/team3/Git/cse284_project/Data >> /home/hmummey/teams/CSE284_SP21_A00/team3/HMM_Test_Outputs/simulated_files.txt\n",
    "ls simAd*.tsv >> /home/hmummey/teams/CSE284_SP21_A00/team3/HMM_Test_Outputs/simulated_files.txt\n",
    "ls simGen*.tsv >> /home/hmummey/teams/CSE284_SP21_A00/team3/HMM_Test_Outputs/simulated_files.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "contrary-blackberry",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/hmummey/teams/CSE284_SP21_A00/team3/Git/cse284_project/Data\n",
      "simAdmixedGenome_100_0_1_Rx1_0.tsv\n",
      "simAdmixedGenome_100_0_1_Rx1_1.tsv\n",
      "simAdmixedGenome_100_0_1_Rx2_0.tsv\n",
      "simAdmixedGenome_100_0_1_Rx2_1.tsv\n",
      "simAdmixedGenome_100_0_1_Rx4_0.tsv\n",
      "simAdmixedGenome_100_0_1_Rx4_1.tsv\n",
      "simAdmixedGenome_100_0_1_Rx5_0.tsv\n",
      "simAdmixedGenome_100_0_1_Rx5_1.tsv\n",
      "simGenome_100_0_0.tsv\n",
      "simGenome_100_0_1.tsv\n",
      "simGenome_100_1_0.tsv\n",
      "simGenome_100_1_1.tsv\n",
      "simGenome_perfect_100_0_0.tsv\n",
      "simGenome_perfect_100_0_1.tsv\n",
      "simGenome_perfect_100_1_0.tsv\n",
      "simGenome_perfect_100_1_1.tsv\n"
     ]
    }
   ],
   "source": [
    "%%bash \n",
    "head -n 20 /home/hmummey/teams/CSE284_SP21_A00/team3/HMM_Test_Outputs/simulated_files.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "enormous-prison",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/hmummey/teams/CSE284_SP21_A00/team3/Git/cse284_project/Data/simAdmixedGenome_100_0_1_Rx1_0.tsv\n",
      "/home/hmummey/teams/CSE284_SP21_A00/team3/Git/cse284_project/Data/simAdmixedGenome_100_0_1_Rx1_1.tsv\n",
      "/home/hmummey/teams/CSE284_SP21_A00/team3/Git/cse284_project/Data/simAdmixedGenome_100_0_1_Rx2_0.tsv\n",
      "/home/hmummey/teams/CSE284_SP21_A00/team3/Git/cse284_project/Data/simAdmixedGenome_100_0_1_Rx2_1.tsv\n",
      "/home/hmummey/teams/CSE284_SP21_A00/team3/Git/cse284_project/Data/simAdmixedGenome_100_0_1_Rx4_0.tsv\n",
      "/home/hmummey/teams/CSE284_SP21_A00/team3/Git/cse284_project/Data/simAdmixedGenome_100_0_1_Rx4_1.tsv\n",
      "/home/hmummey/teams/CSE284_SP21_A00/team3/Git/cse284_project/Data/simAdmixedGenome_100_0_1_Rx5_0.tsv\n",
      "/home/hmummey/teams/CSE284_SP21_A00/team3/Git/cse284_project/Data/simAdmixedGenome_100_0_1_Rx5_1.tsv\n",
      "/home/hmummey/teams/CSE284_SP21_A00/team3/Git/cse284_project/Data/simGenome_100_0_0.tsv\n",
      "/home/hmummey/teams/CSE284_SP21_A00/team3/Git/cse284_project/Data/simGenome_100_0_1.tsv\n",
      "/home/hmummey/teams/CSE284_SP21_A00/team3/Git/cse284_project/Data/simGenome_100_1_0.tsv\n",
      "/home/hmummey/teams/CSE284_SP21_A00/team3/Git/cse284_project/Data/simGenome_100_1_1.tsv\n",
      "/home/hmummey/teams/CSE284_SP21_A00/team3/Git/cse284_project/Data/simGenome_perfect_100_0_0.tsv\n",
      "/home/hmummey/teams/CSE284_SP21_A00/team3/Git/cse284_project/Data/simGenome_perfect_100_0_1.tsv\n",
      "/home/hmummey/teams/CSE284_SP21_A00/team3/Git/cse284_project/Data/simGenome_perfect_100_1_0.tsv\n",
      "/home/hmummey/teams/CSE284_SP21_A00/team3/Git/cse284_project/Data/simGenome_perfect_100_1_1.tsv\n"
     ]
    }
   ],
   "source": [
    "#open the file with the path and names, then loop through and run the HMM and save the output\n",
    "\n",
    "simulated_fp = \"~/teams/CSE284_SP21_A00/team3/HMM_Test_Outputs/simulated_files.txt\"\n",
    "f = open(simulated_fp, 'r')\n",
    "\n",
    "for line in f:\n",
    "    if \"/home\" in line:\n",
    "        direct = line.split(\"\\n\")[0]\n",
    "        #print(direct)\n",
    "    else:\n",
    "        fp = direct + \"/\" + line.split(\"\\n\")[0]\n",
    "        print(fp)\n",
    "        \n",
    "        #testing readin\n",
    "        #test = pd.read_csv(fp,sep='\\t')\n",
    "        #print(test.head(1))\n",
    "        if \"Admixed\" in fp:\n",
    "            for recombr in [0.001,0.01,0.05,0.1,0.2,0.3,0.4,0.5,0.6,0.7,0.8,0.9]:\n",
    "                output_fp = \"~/teams/CSE284_SP21_A00/team3/HMM_Test_Outputs/\" + line.split(\".\")[0] + \"_recomb\"+ str(recombr) + \"_HMMoutput.tsv\"\n",
    "                #print(output_fp)\n",
    "                HMM_implementation_wrapper(test_PopA, test_PopB, fp, test_emission_df, recombr, output_fp, \"21\")\n",
    "    \n",
    "        else:\n",
    "            recombr = 0.1\n",
    "            output_fp = \"~/teams/CSE284_SP21_A00/team3/HMM_Test_Outputs/\" + line.split(\".\")[0] + \"_recomb\"+ str(recombr) + \"_HMMoutput.tsv\"\n",
    "            #print(output_fp)\n",
    "            HMM_implementation_wrapper(test_PopA, test_PopB, fp, test_emission_df, recombr, output_fp, \"21\")\n",
    "    \n",
    "f.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "through-parker",
   "metadata": {},
   "source": [
    "# Medium Length Genomes (5/31/21)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "alive-cleaners",
   "metadata": {},
   "outputs": [],
   "source": [
    "#define names of the two test populations\n",
    "test_PopA = \"0\"\n",
    "test_PopB = \"1\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "funky-hawaiian",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>POS</th>\n",
       "      <th>0_A</th>\n",
       "      <th>0_C</th>\n",
       "      <th>0_G</th>\n",
       "      <th>0_T</th>\n",
       "      <th>1_A</th>\n",
       "      <th>1_C</th>\n",
       "      <th>1_G</th>\n",
       "      <th>1_T</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.556700</td>\n",
       "      <td>0.443300</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.648700</td>\n",
       "      <td>0.351300</td>\n",
       "      <td>0.000001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>0.461900</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.538100</td>\n",
       "      <td>0.288000</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.712000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.842500</td>\n",
       "      <td>0.157500</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.334700</td>\n",
       "      <td>0.665300</td>\n",
       "      <td>0.000001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>0.274400</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.725600</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.954200</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.045800</td>\n",
       "      <td>0.000001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.397200</td>\n",
       "      <td>0.602800</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.492600</td>\n",
       "      <td>0.507400</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   POS       0_A       0_C       0_G       0_T       1_A       1_C       1_G  \\\n",
       "0    0  0.000001  0.556700  0.443300  0.000001  0.000001  0.648700  0.351300   \n",
       "1    1  0.461900  0.000001  0.000001  0.538100  0.288000  0.000001  0.000001   \n",
       "2    2  0.000001  0.842500  0.157500  0.000001  0.000001  0.334700  0.665300   \n",
       "3    3  0.274400  0.000001  0.725600  0.000001  0.954200  0.000001  0.045800   \n",
       "4    4  0.000001  0.000001  0.397200  0.602800  0.000001  0.000001  0.492600   \n",
       "\n",
       "        1_T  \n",
       "0  0.000001  \n",
       "1  0.712000  \n",
       "2  0.000001  \n",
       "3  0.000001  \n",
       "4  0.507400  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#read in emissions data (simulated df)\n",
    "#emission_fp = \"simulated_files/simData_N2_P5000_seed531.tsv\" #assumes in team3 dir\n",
    "emission_fp = \"simulated_files/simData_N2_P5000_seed531_formatted.tsv\"\n",
    "emission_df5000 = pd.read_csv(emission_fp, sep='\\t',header=0)\n",
    "\n",
    "#renaming the columns (only necessary for simulated data)\n",
    "colnames = [\"POS\",\"0_A\",\"0_C\",\"0_G\",\"0_T\",\"1_A\",\"1_C\",\"1_G\",\"1_T\"]\n",
    "emission_df5000.columns = colnames\n",
    "emission_df5000.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "defined-scanner",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "#collect all the file names in one file\n",
    "touch HMM_Test_Outputs/5kSNP_files.txt\n",
    "cd simulated_files\n",
    "\n",
    "echo simulated_files >> ../HMM_Test_Outputs/5kSNP_files.txt\n",
    "ls simAdmixedGenome_5000*.tsv >> ../HMM_Test_Outputs/5kSNP_files.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "failing-ocean",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "simAdmixedGenome_5000_0_1_Rx10_0.tsv\n",
      "simAdmixedGenome_5000_0_1_Rx10_1.tsv\n",
      "simAdmixedGenome_5000_0_1_Rx1_0.tsv\n",
      "simAdmixedGenome_5000_0_1_Rx1_1.tsv\n",
      "simAdmixedGenome_5000_0_1_Rx2_0.tsv\n",
      "simAdmixedGenome_5000_0_1_Rx2_1.tsv\n",
      "simAdmixedGenome_5000_0_1_Rx4_0.tsv\n",
      "simAdmixedGenome_5000_0_1_Rx4_1.tsv\n",
      "simAdmixedGenome_5000_0_1_Rx5_0.tsv\n",
      "simAdmixedGenome_5000_0_1_Rx5_1.tsv\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "# Should only have 10 files\n",
    "head -n 15 HMM_Test_Outputs/5kSNP_files.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "rotary-footage",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "simulated_files/simAdmixedGenome_5000_0_1_Rx10_0.tsv\n",
      "simulated_files/simAdmixedGenome_5000_0_1_Rx10_1.tsv\n",
      "simulated_files/simAdmixedGenome_5000_0_1_Rx1_0.tsv\n",
      "simulated_files/simAdmixedGenome_5000_0_1_Rx1_1.tsv\n",
      "simulated_files/simAdmixedGenome_5000_0_1_Rx2_0.tsv\n",
      "simulated_files/simAdmixedGenome_5000_0_1_Rx2_1.tsv\n",
      "simulated_files/simAdmixedGenome_5000_0_1_Rx4_0.tsv\n",
      "simulated_files/simAdmixedGenome_5000_0_1_Rx4_1.tsv\n",
      "simulated_files/simAdmixedGenome_5000_0_1_Rx5_0.tsv\n",
      "simulated_files/simAdmixedGenome_5000_0_1_Rx5_1.tsv\n"
     ]
    }
   ],
   "source": [
    "#open the file with the path and names, then loop through and run the HMM and save the output\n",
    "\n",
    "simulated_fp = \"HMM_Test_Outputs/5kSNP_files.txt\"\n",
    "f = open(simulated_fp, 'r')\n",
    "\n",
    "for line in f:\n",
    "    if \"/home\" in line:\n",
    "        direct = line.split(\"\\n\")[0]\n",
    "        #print(direct)\n",
    "    else:\n",
    "        #fp = direct + \"/\" + line.split(\"\\n\")[0]\n",
    "        fp=\"simulated_files/\" + line.split(\"\\n\")[0]\n",
    "        print(fp)\n",
    "        \n",
    "        #recomb rates: 1/10000, 1/7500, 1/5000, 1/2500, 1/1000, 1/500, 1/100\n",
    "        #recomb_rates = [0.00013, 0.0001, 0.0002, 0.0004, 0.001, 0.002, 0.01, 0.05, 0.1, 0.2, 0.3, 0.4, 0.5]\n",
    "        #recomb_rates = [0.05, 0.1, 0.2, 0.3, 0.4, 0.5]\n",
    "        recomb_rates = [0.00013, 0.0001, 0.0002, 0.0004, 0.001, 0.002, 0.01]\n",
    "        for recombr in recomb_rates:\n",
    "            #print(str(recombr)[:7])\n",
    "            output_fp = \"HMM_Test_Outputs/\" + line.split(\".\")[0] + \"_recomb\"+ str(recombr)[:7] + \"_HMMoutput.tsv\"\n",
    "            #print(output_fp)\n",
    "            HMM_implementation_wrapper(test_PopA, test_PopB, fp, emission_df5000, recombr, output_fp, \"21\")\n",
    "    \n",
    "f.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "prescription-match",
   "metadata": {},
   "outputs": [],
   "source": [
    "#running it on one Rx4 5k file and getting the emissions matrix\n",
    "\n",
    "#read in sample genotype to test -- assumes current dir is the team3 dir\n",
    "genotype_df = pd.read_csv(\"simulated_files/simAdmixedGenome_5000_0_1_Rx4_1.tsv\", sep='\\t',header = 0)\n",
    "\n",
    "#cut down both dfs to only be positions shared in both\n",
    "fin_emission_df, fin_genotype_df = standardizeIndices(emission_df5000, genotype_df, \"POS\")\n",
    "\n",
    "#create SNPseq\n",
    "a1_idx = list(fin_genotype_df.columns).index('A1')\n",
    "a2_idx = list(fin_genotype_df.columns).index('A2')\n",
    "snps = makeSNPseq(fin_genotype_df,a1_idx,a2_idx)\n",
    "\n",
    "#create the hmm object\n",
    "recomb = 1/5000\n",
    "hmm = HMMOptimalPathLAI(test_PopA, test_PopB, fin_emission_df, recomb, snps)\n",
    "hmm.get_transition_matrix()\n",
    "\n",
    "#perform the Viterbi algorithm and reconstruct the most probable path through states\n",
    "hmm.get_optimal_path()\n",
    "probs_df = hmm.optimal_matrix\n",
    "\n",
    "hmm.reconstruct_path()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "absolute-cleveland",
   "metadata": {},
   "outputs": [],
   "source": [
    "#convert the path to the desired output format (pass the string to use for CHR col)\n",
    "path_df = hmm.output_path(\"21\")\n",
    "#path_df.to_csv(output_fp,sep='\\t',header=True,index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "swedish-slide",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    4.006872e-321\n",
       "1    9.896542e-317\n",
       "2     0.000000e+00\n",
       "3    1.403146e-321\n",
       "Name: 735, dtype: float64"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "probs_df2 = pd.DataFrame(probs_df)\n",
    "probs_df2[735]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "owned-reasoning",
   "metadata": {},
   "outputs": [],
   "source": [
    "probs_df2.to_csv(\"simAdmixedGenome_5000_0_1_Rx4_1_Probabilities.tsv\",sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "several-election",
   "metadata": {},
   "source": [
    "# Running the HMM on Long Simulated Admixed Genotypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "occasional-gathering",
   "metadata": {},
   "outputs": [],
   "source": [
    "popA = \"AFR\"\n",
    "popB = \"EUR\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "intellectual-cement",
   "metadata": {},
   "source": [
    "Emissions files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "compliant-migration",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHR</th>\n",
       "      <th>POS</th>\n",
       "      <th>SNP</th>\n",
       "      <th>A2</th>\n",
       "      <th>A1</th>\n",
       "      <th>MAF_AFR</th>\n",
       "      <th>MAF_EUR</th>\n",
       "      <th>AFR_A</th>\n",
       "      <th>AFR_C</th>\n",
       "      <th>AFR_G</th>\n",
       "      <th>AFR_T</th>\n",
       "      <th>EUR_A</th>\n",
       "      <th>EUR_C</th>\n",
       "      <th>EUR_G</th>\n",
       "      <th>EUR_T</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>14</td>\n",
       "      <td>19000017</td>\n",
       "      <td>rs375700886</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>14</td>\n",
       "      <td>19000050</td>\n",
       "      <td>rs543746158</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>14</td>\n",
       "      <td>19000056</td>\n",
       "      <td>rs561973970</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>14</td>\n",
       "      <td>19000059</td>\n",
       "      <td>rs201622908</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.00498</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.995020</td>\n",
       "      <td>0.004980</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>14</td>\n",
       "      <td>19000060</td>\n",
       "      <td>rs28973059</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>0.1188</td>\n",
       "      <td>0.38840</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.881200</td>\n",
       "      <td>0.118800</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.611600</td>\n",
       "      <td>0.388400</td>\n",
       "      <td>0.000001</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   CHR       POS          SNP A2 A1  MAF_AFR  MAF_EUR     AFR_A     AFR_C  \\\n",
       "0   14  19000017  rs375700886  C  T   0.0000  0.00000  0.000001  1.000000   \n",
       "1   14  19000050  rs543746158  G  A   0.0000  0.00000  0.000001  0.000001   \n",
       "2   14  19000056  rs561973970  A  T   0.0000  0.00000  1.000000  0.000001   \n",
       "3   14  19000059  rs201622908  G  T   0.0000  0.00498  0.000001  0.000001   \n",
       "4   14  19000060   rs28973059  C  G   0.1188  0.38840  0.000001  0.881200   \n",
       "\n",
       "      AFR_G     AFR_T     EUR_A     EUR_C     EUR_G     EUR_T  \n",
       "0  0.000001  0.000001  0.000001  1.000000  0.000001  0.000001  \n",
       "1  1.000000  0.000001  0.000001  0.000001  1.000000  0.000001  \n",
       "2  0.000001  0.000001  1.000000  0.000001  0.000001  0.000001  \n",
       "3  1.000000  0.000001  0.000001  0.000001  0.995020  0.004980  \n",
       "4  0.118800  0.000001  0.000001  0.611600  0.388400  0.000001  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#chr14\n",
    "chr14_emissions_fp = \"/home/hmummey/teams/CSE284_SP21_A00/team3/chromosome_14_files/chr14_genotypes_afr_eur_allelefreqs.bybp.csv\"\n",
    "emission_df_chr14 = pd.read_csv(chr14_emissions_fp, sep=',',header=0)\n",
    "emission_df_chr14.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "continuing-smile",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHR</th>\n",
       "      <th>POS</th>\n",
       "      <th>SNP</th>\n",
       "      <th>A2</th>\n",
       "      <th>A1</th>\n",
       "      <th>MAF_AFR</th>\n",
       "      <th>MAF_EUR</th>\n",
       "      <th>AFR_A</th>\n",
       "      <th>AFR_C</th>\n",
       "      <th>AFR_G</th>\n",
       "      <th>AFR_T</th>\n",
       "      <th>EUR_A</th>\n",
       "      <th>EUR_C</th>\n",
       "      <th>EUR_G</th>\n",
       "      <th>EUR_T</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>21</td>\n",
       "      <td>9411239</td>\n",
       "      <td>rs559462325</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>21</td>\n",
       "      <td>9411245</td>\n",
       "      <td>rs181691356</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>0.000893</td>\n",
       "      <td>0.001992</td>\n",
       "      <td>0.000893</td>\n",
       "      <td>0.999107</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.001992</td>\n",
       "      <td>0.998008</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>21</td>\n",
       "      <td>9411264</td>\n",
       "      <td>rs548263598</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>0.000893</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.999107</td>\n",
       "      <td>0.000893</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>21</td>\n",
       "      <td>9411267</td>\n",
       "      <td>rs561987868</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>21</td>\n",
       "      <td>9411302</td>\n",
       "      <td>rs531010746</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>0.011610</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.988390</td>\n",
       "      <td>0.011610</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000001</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   CHR      POS          SNP A2 A1   MAF_AFR   MAF_EUR     AFR_A     AFR_C  \\\n",
       "0   21  9411239  rs559462325  G  A  0.000000  0.000000  0.000001  0.000001   \n",
       "1   21  9411245  rs181691356  C  A  0.000893  0.001992  0.000893  0.999107   \n",
       "2   21  9411264  rs548263598  A  C  0.000893  0.000000  0.999107  0.000893   \n",
       "3   21  9411267  rs561987868  G  T  0.000000  0.000000  0.000001  0.000001   \n",
       "4   21  9411302  rs531010746  G  T  0.011610  0.000000  0.000001  0.000001   \n",
       "\n",
       "      AFR_G     AFR_T     EUR_A     EUR_C     EUR_G     EUR_T  \n",
       "0  1.000000  0.000001  0.000001  0.000001  1.000000  0.000001  \n",
       "1  0.000001  0.000001  0.001992  0.998008  0.000001  0.000001  \n",
       "2  0.000001  0.000001  1.000000  0.000001  0.000001  0.000001  \n",
       "3  1.000000  0.000001  0.000001  0.000001  1.000000  0.000001  \n",
       "4  0.988390  0.011610  0.000001  0.000001  1.000000  0.000001  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#chr21\n",
    "chr21_emissions_fp = \"/home/hmummey/teams/CSE284_SP21_A00/team3/chromosome_21_files/chr21_genotypes_afr_eur_allelefreqs.bybp.csv\"\n",
    "emission_df_chr21 = pd.read_csv(chr21_emissions_fp, sep=',',header=0)\n",
    "emission_df_chr21.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "integrated-hobby",
   "metadata": {},
   "source": [
    "Genotype files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "parliamentary-conducting",
   "metadata": {},
   "outputs": [],
   "source": [
    "# %%bash\n",
    "\n",
    "# #collect all the file names in one file\n",
    "# touch /home/hmummey/teams/CSE284_SP21_A00/team3/HMM_Test_Outputs/admixed_files.txt\n",
    "# cd /home/hmummey/teams/CSE284_SP21_A00/team3/simulated_files\n",
    "\n",
    "# echo /home/hmummey/teams/CSE284_SP21_A00/team3/simulated_files >> /home/hmummey/teams/CSE284_SP21_A00/team3/HMM_Test_Outputs/admixed_files.txt\n",
    "# ls admix*.tsv >> /home/hmummey/teams/CSE284_SP21_A00/team3/HMM_Test_Outputs/admixed_files.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "mediterranean-gambling",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/hmummey/teams/CSE284_SP21_A00/team3/simulated_files\n",
      "admixEUR_AFR_chr14_Rx1_a.tsv\n",
      "admixEUR_AFR_chr14_Rx1_b.tsv\n",
      "admixEUR_AFR_chr14_Rx1_c.tsv\n",
      "admixEUR_AFR_chr14_Rx3_a.tsv\n",
      "admixEUR_AFR_chr14_Rx3_b.tsv\n",
      "admixEUR_AFR_chr14_Rx3_c.tsv\n",
      "admixEUR_AFR_chr21_Rx1_a.tsv\n",
      "admixEUR_AFR_chr21_Rx1_b.tsv\n",
      "admixEUR_AFR_chr21_Rx1_c.tsv\n",
      "admixEUR_AFR_chr21_Rx3_a.tsv\n",
      "admixEUR_AFR_chr21_Rx3_b.tsv\n",
      "admixEUR_AFR_chr21_Rx3_c.tsv\n",
      "admixEUR_Rx1a_chr21_Rx1_a.tsv\n",
      "admixEUR_Rx1a_chr21_Rx1_b.tsv\n",
      "admixEUR_Rx1a_chr21_Rx1_c.tsv\n",
      "admixRx1c_Rx3b_chr21_Rx1_a.tsv\n",
      "admixRx1c_Rx3b_chr21_Rx1_b.tsv\n",
      "admixRx1c_Rx3b_chr21_Rx1_c.tsv\n"
     ]
    }
   ],
   "source": [
    "%%bash \n",
    "head -n 20 /home/hmummey/teams/CSE284_SP21_A00/team3/HMM_Test_Outputs/admixed_files.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "controversial-credit",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/hmummey/teams/CSE284_SP21_A00/team3/simulated_files/admixEUR_AFR_chr14_Rx1_a.tsv\n",
      "2021-05-26 20:18:00\n"
     ]
    },
    {
     "ename": "KeyError",
     "evalue": "682",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[0;32m/opt/conda/lib/python3.8/site-packages/pandas/core/indexes/base.py\u001b[0m in \u001b[0;36mget_loc\u001b[0;34m(self, key, method, tolerance)\u001b[0m\n\u001b[1;32m   3079\u001b[0m             \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 3080\u001b[0;31m                 \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_engine\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_loc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcasted_key\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   3081\u001b[0m             \u001b[0;32mexcept\u001b[0m \u001b[0mKeyError\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0merr\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/index.pyx\u001b[0m in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/index.pyx\u001b[0m in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: 682",
      "\nThe above exception was the direct cause of the following exception:\n",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-69-4e41459fdddc>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     22\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     23\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0;34m\"14\"\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mfp\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 24\u001b[0;31m             \u001b[0mHMM_implementation_wrapper\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mpopA\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mpopB\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfp\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0memission_df_chr14\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m0.1\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0moutput_fp\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"14\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     25\u001b[0m         \u001b[0;32melif\u001b[0m \u001b[0;34m\"21\"\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mfp\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     26\u001b[0m             \u001b[0mHMM_implementation_wrapper\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mpopA\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mpopB\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfp\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0memission_df_chr21\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m0.1\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0moutput_fp\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"21\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<ipython-input-65-a8a355cec0bb>\u001b[0m in \u001b[0;36mHMM_implementation_wrapper\u001b[0;34m(popA, popB, genotype_fp, emission_df, recomb, output_fp, chrm)\u001b[0m\n\u001b[1;32m     16\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     17\u001b[0m     \u001b[0;31m#perform the Viterbi algorithm and reconstruct the most probable path through states\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 18\u001b[0;31m     \u001b[0mhmm\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_optimal_path\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     19\u001b[0m     \u001b[0mhmm\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreconstruct_path\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     20\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/teams/CSE284_SP21_A00/team3/LAI_hmm_scriptFINAL.py\u001b[0m in \u001b[0;36mget_optimal_path\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    135\u001b[0m             \u001b[0midx\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0malphabet\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mindex\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mc\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    136\u001b[0m             \u001b[0;31m#create the emissions matrix for this SNP with the enumerated index\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 137\u001b[0;31m             \u001b[0msnp_emissions\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_emissions_matrix\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mi\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    138\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    139\u001b[0m             \u001b[0;31m#go through all states and calculate max probability based on emission and transition probs\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/teams/CSE284_SP21_A00/team3/LAI_hmm_scriptFINAL.py\u001b[0m in \u001b[0;36mget_emissions_matrix\u001b[0;34m(self, ind)\u001b[0m\n\u001b[1;32m    115\u001b[0m                 \u001b[0mnt1\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0malphabet\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mj\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    116\u001b[0m                 \u001b[0mnt2\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0malphabet\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mj\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 117\u001b[0;31m                 \u001b[0mprob1\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0memissions_df\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mloc\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mind\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mpop1\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0;34m\"_\"\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0mnt1\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    118\u001b[0m                 \u001b[0mprob2\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0memissions_df\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mloc\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mind\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mpop2\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0;34m\"_\"\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0mnt2\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    119\u001b[0m                 \u001b[0memissions_mat\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mi\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mj\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mprob1\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0mprob2\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/conda/lib/python3.8/site-packages/pandas/core/indexing.py\u001b[0m in \u001b[0;36m__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m    886\u001b[0m                     \u001b[0;31m# AttributeError for IntervalTree get_value\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    887\u001b[0m                     \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mobj\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_get_value\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtakeable\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_takeable\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 888\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_getitem_tuple\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    889\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    890\u001b[0m             \u001b[0;31m# we by definition only have the 0th axis\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/conda/lib/python3.8/site-packages/pandas/core/indexing.py\u001b[0m in \u001b[0;36m_getitem_tuple\u001b[0;34m(self, tup)\u001b[0m\n\u001b[1;32m   1057\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m_getitem_tuple\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtup\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mTuple\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1058\u001b[0m         \u001b[0;32mwith\u001b[0m \u001b[0msuppress\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mIndexingError\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1059\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_getitem_lowerdim\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtup\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1060\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1061\u001b[0m         \u001b[0;31m# no multi-index, so validate all of the indexers\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/conda/lib/python3.8/site-packages/pandas/core/indexing.py\u001b[0m in \u001b[0;36m_getitem_lowerdim\u001b[0;34m(self, tup)\u001b[0m\n\u001b[1;32m    804\u001b[0m                 \u001b[0;31m# We don't need to check for tuples here because those are\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    805\u001b[0m                 \u001b[0;31m#  caught by the _is_nested_tuple_indexer check above.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 806\u001b[0;31m                 \u001b[0msection\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_getitem_axis\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mi\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    807\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    808\u001b[0m                 \u001b[0;31m# We should never have a scalar section here, because\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/conda/lib/python3.8/site-packages/pandas/core/indexing.py\u001b[0m in \u001b[0;36m_getitem_axis\u001b[0;34m(self, key, axis)\u001b[0m\n\u001b[1;32m   1121\u001b[0m         \u001b[0;31m# fall thru to straight lookup\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1122\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_validate_key\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1123\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_get_label\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0maxis\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1124\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1125\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m_get_slice_axis\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mslice_obj\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mslice\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mint\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/conda/lib/python3.8/site-packages/pandas/core/indexing.py\u001b[0m in \u001b[0;36m_get_label\u001b[0;34m(self, label, axis)\u001b[0m\n\u001b[1;32m   1070\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m_get_label\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mlabel\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mint\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1071\u001b[0m         \u001b[0;31m# GH#5667 this will fail if the label is not present in the axis.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1072\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mobj\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mxs\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlabel\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0maxis\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1073\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1074\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m_handle_lowerdim_multi_index_axis0\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtup\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mTuple\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/conda/lib/python3.8/site-packages/pandas/core/generic.py\u001b[0m in \u001b[0;36mxs\u001b[0;34m(self, key, axis, level, drop_level)\u001b[0m\n\u001b[1;32m   3734\u001b[0m                 \u001b[0;32mraise\u001b[0m \u001b[0mTypeError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34mf\"Expected label or tuple of labels, got {key}\"\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0me\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3735\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 3736\u001b[0;31m             \u001b[0mloc\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mindex\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_loc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   3737\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3738\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0misinstance\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mloc\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mndarray\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/conda/lib/python3.8/site-packages/pandas/core/indexes/base.py\u001b[0m in \u001b[0;36mget_loc\u001b[0;34m(self, key, method, tolerance)\u001b[0m\n\u001b[1;32m   3080\u001b[0m                 \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_engine\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_loc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcasted_key\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3081\u001b[0m             \u001b[0;32mexcept\u001b[0m \u001b[0mKeyError\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0merr\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 3082\u001b[0;31m                 \u001b[0;32mraise\u001b[0m \u001b[0mKeyError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0merr\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   3083\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3084\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mtolerance\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: 682"
     ]
    }
   ],
   "source": [
    "# #open the file with the path and names, then loop through and run the HMM and save the output\n",
    "\n",
    "# admixed_fp = \"/home/hmummey/teams/CSE284_SP21_A00/team3/HMM_Test_Outputs/admixed_files.txt\"\n",
    "# f = open(admixed_fp, 'r')\n",
    "\n",
    "# for line in f:\n",
    "#     if \"/home\" in line:\n",
    "#         direct = line.split(\"\\n\")[0]\n",
    "#         #print(direct)\n",
    "#     else:\n",
    "#         fp = direct + \"/\" + line.split(\"\\n\")[0]\n",
    "#         print(fp)\n",
    "#         now = datetime.datetime.now()\n",
    "#         print (now.strftime(\"%Y-%m-%d %H:%M:%S\"))\n",
    "        \n",
    "#         #testing readin\n",
    "#         #test = pd.read_csv(fp,sep='\\t')\n",
    "#         #print(test.head(1))\n",
    "        \n",
    "#         output_fp = \"/home/hmummey/teams/CSE284_SP21_A00/team3/HMM_Test_Outputs/\" + line.split(\".\")[0] + \"_HMMoutput.tsv\"\n",
    "#         #print(output_fp)\n",
    "        \n",
    "#         if \"14\" in fp:\n",
    "#             HMM_implementation_wrapper(popA, popB, fp, emission_df_chr14, 0.1, output_fp, \"14\")\n",
    "#         elif \"21\" in fp:\n",
    "#             HMM_implementation_wrapper(popA, popB, fp, emission_df_chr21, 0.1, output_fp, \"21\")\n",
    "        \n",
    "    \n",
    "# f.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "comprehensive-surprise",
   "metadata": {},
   "source": [
    "# Troubleshooting on Long Simulated Admixed (one genome)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "anonymous-foundation",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2505787\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHR</th>\n",
       "      <th>POS</th>\n",
       "      <th>SNP</th>\n",
       "      <th>A2</th>\n",
       "      <th>A1</th>\n",
       "      <th>MAF_AFR</th>\n",
       "      <th>MAF_EUR</th>\n",
       "      <th>AFR_A</th>\n",
       "      <th>AFR_C</th>\n",
       "      <th>AFR_G</th>\n",
       "      <th>AFR_T</th>\n",
       "      <th>EUR_A</th>\n",
       "      <th>EUR_C</th>\n",
       "      <th>EUR_G</th>\n",
       "      <th>EUR_T</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>14</td>\n",
       "      <td>19000017</td>\n",
       "      <td>rs375700886</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>14</td>\n",
       "      <td>19000050</td>\n",
       "      <td>rs543746158</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>14</td>\n",
       "      <td>19000056</td>\n",
       "      <td>rs561973970</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>14</td>\n",
       "      <td>19000059</td>\n",
       "      <td>rs201622908</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.00498</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.995020</td>\n",
       "      <td>0.004980</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>14</td>\n",
       "      <td>19000060</td>\n",
       "      <td>rs28973059</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>0.1188</td>\n",
       "      <td>0.38840</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.881200</td>\n",
       "      <td>0.118800</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.611600</td>\n",
       "      <td>0.388400</td>\n",
       "      <td>0.000001</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   CHR       POS          SNP A2 A1  MAF_AFR  MAF_EUR     AFR_A     AFR_C  \\\n",
       "0   14  19000017  rs375700886  C  T   0.0000  0.00000  0.000001  1.000000   \n",
       "1   14  19000050  rs543746158  G  A   0.0000  0.00000  0.000001  0.000001   \n",
       "2   14  19000056  rs561973970  A  T   0.0000  0.00000  1.000000  0.000001   \n",
       "3   14  19000059  rs201622908  G  T   0.0000  0.00498  0.000001  0.000001   \n",
       "4   14  19000060   rs28973059  C  G   0.1188  0.38840  0.000001  0.881200   \n",
       "\n",
       "      AFR_G     AFR_T     EUR_A     EUR_C     EUR_G     EUR_T  \n",
       "0  0.000001  0.000001  0.000001  1.000000  0.000001  0.000001  \n",
       "1  1.000000  0.000001  0.000001  0.000001  1.000000  0.000001  \n",
       "2  0.000001  0.000001  1.000000  0.000001  0.000001  0.000001  \n",
       "3  1.000000  0.000001  0.000001  0.000001  0.995020  0.004980  \n",
       "4  0.118800  0.000001  0.000001  0.611600  0.388400  0.000001  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(len(emission_df_chr14))\n",
    "emission_df_chr14.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "pretty-flour",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2539145\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>POS</th>\n",
       "      <th>A1</th>\n",
       "      <th>A2</th>\n",
       "      <th>POP1</th>\n",
       "      <th>POP2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>19000017</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>19000050</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>19000056</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>19000059</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>19000060</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        POS A1 A2  POP1  POP2\n",
       "0  19000017  T  T     1     0\n",
       "1  19000050  A  A     1     0\n",
       "2  19000056  T  T     1     0\n",
       "3  19000059  T  T     1     0\n",
       "4  19000060  G  G     1     0"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#go through manually (step by step) for the first file: /home/hmummey/teams/CSE284_SP21_A00/team3/simulated_files/admixEUR_AFR_chr14_Rx1_a.tsv\n",
    "admixed_fp1 = \"/home/hmummey/teams/CSE284_SP21_A00/team3/simulated_files/admixEUR_AFR_chr14_Rx1_a.tsv\"\n",
    "genotype_df2 = pd.read_csv(admixed_fp1, sep='\\t',header = 0)\n",
    "print(len(genotype_df2))\n",
    "genotype_df2.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "egyptian-packing",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2500619\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHR</th>\n",
       "      <th>POS</th>\n",
       "      <th>SNP</th>\n",
       "      <th>A2</th>\n",
       "      <th>A1</th>\n",
       "      <th>MAF_AFR</th>\n",
       "      <th>MAF_EUR</th>\n",
       "      <th>AFR_A</th>\n",
       "      <th>AFR_C</th>\n",
       "      <th>AFR_G</th>\n",
       "      <th>AFR_T</th>\n",
       "      <th>EUR_A</th>\n",
       "      <th>EUR_C</th>\n",
       "      <th>EUR_G</th>\n",
       "      <th>EUR_T</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>14</td>\n",
       "      <td>19000017</td>\n",
       "      <td>rs375700886</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>14</td>\n",
       "      <td>19000050</td>\n",
       "      <td>rs543746158</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>14</td>\n",
       "      <td>19000056</td>\n",
       "      <td>rs561973970</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>14</td>\n",
       "      <td>19000059</td>\n",
       "      <td>rs201622908</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.00498</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.995020</td>\n",
       "      <td>0.004980</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>14</td>\n",
       "      <td>19000060</td>\n",
       "      <td>rs28973059</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>0.1188</td>\n",
       "      <td>0.38840</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.881200</td>\n",
       "      <td>0.118800</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.611600</td>\n",
       "      <td>0.388400</td>\n",
       "      <td>0.000001</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   CHR       POS          SNP A2 A1  MAF_AFR  MAF_EUR     AFR_A     AFR_C  \\\n",
       "0   14  19000017  rs375700886  C  T   0.0000  0.00000  0.000001  1.000000   \n",
       "1   14  19000050  rs543746158  G  A   0.0000  0.00000  0.000001  0.000001   \n",
       "2   14  19000056  rs561973970  A  T   0.0000  0.00000  1.000000  0.000001   \n",
       "3   14  19000059  rs201622908  G  T   0.0000  0.00498  0.000001  0.000001   \n",
       "4   14  19000060   rs28973059  C  G   0.1188  0.38840  0.000001  0.881200   \n",
       "\n",
       "      AFR_G     AFR_T     EUR_A     EUR_C     EUR_G     EUR_T  \n",
       "0  0.000001  0.000001  0.000001  1.000000  0.000001  0.000001  \n",
       "1  1.000000  0.000001  0.000001  0.000001  1.000000  0.000001  \n",
       "2  0.000001  0.000001  1.000000  0.000001  0.000001  0.000001  \n",
       "3  1.000000  0.000001  0.000001  0.000001  0.995020  0.004980  \n",
       "4  0.118800  0.000001  0.000001  0.611600  0.388400  0.000001  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fin_emission_df2, fin_genotype_df2 = standardizeIndices(emission_df_chr14, genotype_df2, \"POS\")\n",
    "print(len(fin_emission_df2))\n",
    "fin_emission_df2.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "going-newman",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2500619\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>POS</th>\n",
       "      <th>A1</th>\n",
       "      <th>A2</th>\n",
       "      <th>POP1</th>\n",
       "      <th>POP2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>19000017</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>19000050</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>19000056</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>19000059</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>19000060</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        POS A1 A2  POP1  POP2\n",
       "0  19000017  T  T     1     0\n",
       "1  19000050  A  A     1     0\n",
       "2  19000056  T  T     1     0\n",
       "3  19000059  T  T     1     0\n",
       "4  19000060  G  G     1     0"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(len(fin_genotype_df2))\n",
    "fin_genotype_df2.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "automated-corrections",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1 2\n",
      "2021-05-27 11:45:16\n",
      "['TT', 'AA', 'TT', 'TT', 'GG', 'AA', 'CC', 'GG', 'GG', 'TT']\n",
      "2021-05-27 11:47:33\n"
     ]
    }
   ],
   "source": [
    "#create SNPseq\n",
    "a1_idx = list(fin_genotype_df2.columns).index('A1')\n",
    "a2_idx = list(fin_genotype_df2.columns).index('A2')\n",
    "print(a1_idx,a2_idx)\n",
    "now = datetime.datetime.now()\n",
    "print (now.strftime(\"%Y-%m-%d %H:%M:%S\"))\n",
    "\n",
    "snps = makeSNPseq(fin_genotype_df2,a1_idx,a2_idx)\n",
    "print(snps[:10])\n",
    "now = datetime.datetime.now()\n",
    "print (now.strftime(\"%Y-%m-%d %H:%M:%S\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "aerial-acrylic",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2500619\n"
     ]
    }
   ],
   "source": [
    "print(len(snps))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "violent-genealogy",
   "metadata": {},
   "outputs": [],
   "source": [
    "recomb = 0.1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "printable-assets",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2021-05-27 11:47:33\n",
      "2021-05-27 11:47:33\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-16-933f68440a24>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      9\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     10\u001b[0m \u001b[0;31m#perform the Viterbi algorithm and reconstruct the most probable path through states\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 11\u001b[0;31m \u001b[0mhmm\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_optimal_path\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     12\u001b[0m \u001b[0mhmm\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreconstruct_path\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     13\u001b[0m \u001b[0mnow\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdatetime\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdatetime\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mnow\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/teams/CSE284_SP21_A00/team3/LAI_hmm_scriptFINAL.py\u001b[0m in \u001b[0;36mget_optimal_path\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    139\u001b[0m             \u001b[0midx\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0malphabet\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mindex\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mc\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    140\u001b[0m             \u001b[0;31m#create the emissions matrix for this SNP with the enumerated index\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 141\u001b[0;31m             \u001b[0msnp_emissions\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_emissions_matrix\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mi\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    142\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    143\u001b[0m             \u001b[0;31m#go through all states and calculate max probability based on emission and transition probs\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/teams/CSE284_SP21_A00/team3/LAI_hmm_scriptFINAL.py\u001b[0m in \u001b[0;36mget_emissions_matrix\u001b[0;34m(self, ind)\u001b[0m\n\u001b[1;32m    119\u001b[0m                 \u001b[0mcol1\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mlist\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0memissions_df\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcolumns\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mindex\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mpop1\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0;34m\"_\"\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0mnt1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    120\u001b[0m                 \u001b[0mcol2\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mlist\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0memissions_df\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcolumns\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mindex\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mpop1\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0;34m\"_\"\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0mnt2\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 121\u001b[0;31m                 \u001b[0mprob1\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0memissions_df\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0miloc\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mind\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mcol1\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    122\u001b[0m                 \u001b[0mprob2\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0memissions_df\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0miloc\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mind\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mcol2\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    123\u001b[0m                 \u001b[0memissions_mat\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mi\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mj\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mprob1\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0mprob2\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/conda/lib/python3.8/site-packages/pandas/core/indexing.py\u001b[0m in \u001b[0;36m__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m    885\u001b[0m                 \u001b[0;32mwith\u001b[0m \u001b[0msuppress\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mKeyError\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mIndexError\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mAttributeError\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    886\u001b[0m                     \u001b[0;31m# AttributeError for IntervalTree get_value\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 887\u001b[0;31m                     \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mobj\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_get_value\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtakeable\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_takeable\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    888\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_getitem_tuple\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    889\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/conda/lib/python3.8/site-packages/pandas/core/frame.py\u001b[0m in \u001b[0;36m_get_value\u001b[0;34m(self, index, col, takeable)\u001b[0m\n\u001b[1;32m   3124\u001b[0m         \"\"\"\n\u001b[1;32m   3125\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mtakeable\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 3126\u001b[0;31m             \u001b[0mseries\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_ixs\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcol\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   3127\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0mseries\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_values\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mindex\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3128\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/conda/lib/python3.8/site-packages/pandas/core/frame.py\u001b[0m in \u001b[0;36m_ixs\u001b[0;34m(self, i, axis)\u001b[0m\n\u001b[1;32m   2963\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2964\u001b[0m             \u001b[0mvalues\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_mgr\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0miget\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mi\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 2965\u001b[0;31m             \u001b[0mresult\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_box_col_values\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mvalues\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mi\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   2966\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2967\u001b[0m             \u001b[0;31m# this is a cached value, mark it so\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/conda/lib/python3.8/site-packages/pandas/core/frame.py\u001b[0m in \u001b[0;36m_box_col_values\u001b[0;34m(self, values, loc)\u001b[0m\n\u001b[1;32m   3309\u001b[0m         \u001b[0mname\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcolumns\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mloc\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3310\u001b[0m         \u001b[0mklass\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_constructor_sliced\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 3311\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mklass\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mvalues\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mindex\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mindex\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfastpath\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   3312\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3313\u001b[0m     \u001b[0;31m# ----------------------------------------------------------------------\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/conda/lib/python3.8/site-packages/pandas/core/series.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, data, index, dtype, name, copy, fastpath)\u001b[0m\n\u001b[1;32m    337\u001b[0m                 \u001b[0mdata\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mSingleBlockManager\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfrom_array\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdata\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mindex\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    338\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 339\u001b[0;31m         \u001b[0mgeneric\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mNDFrame\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__init__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdata\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    340\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mname\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    341\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_set_axis\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mindex\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfastpath\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/conda/lib/python3.8/site-packages/pandas/core/generic.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, data, copy, attrs)\u001b[0m\n\u001b[1;32m    199\u001b[0m             \u001b[0mattrs\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdict\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mattrs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    200\u001b[0m         \u001b[0mobject\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__setattr__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"_attrs\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mattrs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 201\u001b[0;31m         \u001b[0mobject\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__setattr__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"_flags\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mFlags\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mallows_duplicate_labels\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    202\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    203\u001b[0m     \u001b[0;34m@\u001b[0m\u001b[0mclassmethod\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/conda/lib/python3.8/site-packages/pandas/core/flags.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, obj, allows_duplicate_labels)\u001b[0m\n\u001b[1;32m     47\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m__init__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mobj\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m*\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mallows_duplicate_labels\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     48\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_allows_duplicate_labels\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mallows_duplicate_labels\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 49\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_obj\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mweakref\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mref\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mobj\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     50\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     51\u001b[0m     \u001b[0;34m@\u001b[0m\u001b[0mproperty\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "#create the hmm object\n",
    "now = datetime.datetime.now()\n",
    "print (now.strftime(\"%Y-%m-%d %H:%M:%S\"))\n",
    "\n",
    "hmm = HMMOptimalPathLAI(popA, popB, fin_emission_df2, recomb, snps)\n",
    "hmm.get_transition_matrix()\n",
    "now = datetime.datetime.now()\n",
    "print (now.strftime(\"%Y-%m-%d %H:%M:%S\"))\n",
    "\n",
    "#perform the Viterbi algorithm and reconstruct the most probable path through states\n",
    "hmm.get_optimal_path()\n",
    "hmm.reconstruct_path()\n",
    "now = datetime.datetime.now()\n",
    "print (now.strftime(\"%Y-%m-%d %H:%M:%S\"))\n",
    "\n",
    "#convert the path to the desired output format (pass the string to use for CHR col)\n",
    "path_df = hmm.output_path(\"14\")\n",
    "path_df.to_csv(output_fp,sep='\\t',header=True,index=False)\n",
    "now = datetime.datetime.now()\n",
    "print (now.strftime(\"%Y-%m-%d %H:%M:%S\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "mobile-geneva",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "appreciated-sacrifice",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "liquid-costs",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "still-emperor",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "married-action",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dress-loading",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dominant-trading",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "formed-flush",
   "metadata": {},
   "source": [
    "# Debugging"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "square-bacteria",
   "metadata": {},
   "source": [
    "Figuring out weird indexing issue"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "crazy-heather",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "False\n",
      "[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 97, 98, 99, 100, 101, 102, 103, 104, 105, 106, 107, 108, 109, 110, 111, 112, 113, 114, 115, 116, 117, 118, 119, 120, 121, 122, 123, 124, 125, 126, 127, 128, 129, 130, 131, 132, 133, 134, 135, 136, 137, 138, 139, 140, 141, 142, 143, 144, 145, 146, 147, 148, 149, 150, 151, 152, 153, 154, 155, 156, 157, 158, 159, 160, 161, 162, 163, 164, 165, 166, 167, 168, 169, 170, 171, 172, 173, 174, 175, 176, 177, 178, 179, 180, 181, 182, 183, 184, 185, 186, 187, 188, 189, 190, 191, 192, 193, 194, 195, 196, 197, 198, 199, 200, 201, 202, 203, 204, 205, 206, 207, 208, 209, 210, 211, 212, 213, 214, 215, 216, 217, 218, 219, 220, 221, 222, 223, 224, 225, 226, 227, 228, 229, 230, 231, 232, 233, 234, 235, 236, 237, 238, 239, 240, 241, 242, 243, 244, 245, 246, 247, 248, 249, 250, 251, 252, 253, 254, 255, 256, 257, 258, 259, 260, 261, 262, 263, 264, 265, 266, 267, 268, 269, 270, 271, 272, 273, 274, 275, 276, 277, 278, 279, 280, 281, 282, 283, 284, 285, 286, 287, 288, 289, 290, 291, 292, 293, 294, 295, 296, 297, 298, 299, 300, 301, 302, 303, 304, 305, 306, 307, 308, 309, 310, 311, 312, 313, 314, 315, 316, 317, 318, 319, 320, 321, 322, 323, 324, 325, 326, 327, 328, 329, 330, 331, 332, 333, 334, 335, 336, 337, 338, 339, 340, 341, 342, 343, 344, 345, 346, 347, 348, 349, 350, 351, 352, 353, 354, 355, 356, 357, 358, 359, 360, 361, 362, 363, 364, 365, 366, 367, 368, 369, 370, 371, 372, 373, 374, 375, 376, 377, 378, 379, 380, 381, 382, 383, 384, 385, 386, 387, 388, 389, 391, 392, 393, 394, 395, 396, 397, 398, 399, 400, 401, 402, 403, 404, 405, 406, 407, 408, 409, 410, 411, 412, 413, 414, 415, 416, 417, 418, 419, 420, 421, 422, 423, 424, 425, 426, 427, 428, 429, 430, 431, 432, 433, 434, 435, 436, 437, 438, 439, 440, 441, 442, 443, 444, 445, 446, 447, 448, 449, 450, 451, 452, 453, 454, 455, 456, 457, 458, 459, 460, 461, 462, 463, 464, 465, 466, 467, 468, 469, 470, 471, 472, 473, 474, 475, 476, 477, 478, 479, 480, 481, 482, 483, 484, 485, 486, 487, 488, 489, 490, 491, 492, 493, 494, 495, 496, 497, 498, 499, 500, 501, 502, 503, 504, 505, 506, 507, 508, 509, 510, 511, 512, 513, 514, 515, 516, 517, 518, 519, 520, 521, 522, 523, 524, 525, 526, 527, 528, 529, 530, 531, 532, 533, 534, 535, 536, 537, 538, 539, 540, 541, 542, 543, 544, 545, 546, 547, 548, 549, 550, 551, 552, 553, 554, 555, 556, 557, 558, 559, 560, 561, 562, 563, 564, 565, 566, 567, 568, 569, 570, 571, 572, 573, 574, 575, 576, 577, 578, 579, 580, 581, 582, 583, 584, 585, 586, 587, 588, 589, 590, 591, 592, 593, 594, 595, 596, 597, 598, 599, 600, 601, 602, 603, 604, 605, 606, 607, 608, 609, 610, 611, 612, 613, 614, 615, 616, 617, 618, 619, 620, 621, 622, 623, 624, 625, 626, 627, 628, 629, 630, 631, 632, 633, 634, 635, 636, 637, 638, 639, 640, 641, 642, 643, 644, 645, 646, 647, 648, 649, 650, 651, 652, 653, 654, 655, 656, 657, 658, 659, 660, 661, 662, 663, 664, 665, 666, 667, 668, 669, 670, 671, 672, 673, 674, 675, 676, 677, 678, 679, 680, 681, 682, 683, 684, 685, 686, 687, 688, 689, 690, 691, 692, 693, 694, 695, 696, 697, 698, 699, 700, 701, 702, 703, 704, 705, 706, 707, 708, 709, 710, 711, 712, 713, 714, 715, 716, 717, 718, 719, 720, 721, 722, 723, 724, 725, 726, 727, 728, 729, 730, 731, 732, 733, 734, 735, 736, 737, 738, 739, 740, 741, 742, 743, 744, 745, 746, 747, 748, 749, 750, 751, 752, 753, 754, 755, 756, 757, 758, 759, 760, 761, 762, 763, 764, 765, 766, 767, 768, 769, 770, 771, 772, 773, 774, 775, 776, 777, 778, 779, 780, 781, 782, 783, 784, 785, 786, 787, 788, 789, 790, 791, 792, 793, 794, 795, 796, 797, 798, 799, 800, 801, 802, 803, 804, 805, 806, 807, 808, 809, 810, 811, 812, 813, 814, 815, 816, 817, 818, 819, 820, 821, 822, 823, 824, 825, 826, 827, 828, 829, 830, 831, 832, 833, 834, 835, 836, 837, 838, 839, 840, 841, 842, 843, 844, 845, 846, 847, 848, 849, 850, 851, 852, 853, 854, 855, 856, 857, 858, 859, 860, 861, 862, 863, 864, 865, 866, 867, 868, 869, 870, 871, 872, 873, 874, 875, 876, 877, 878, 879, 880, 881, 882, 883, 884, 885, 886, 887, 888, 889, 890, 891, 892, 893, 894, 895, 896, 897, 898, 899, 900, 901, 902, 903, 904, 905, 906, 907, 909, 910, 911, 912, 913, 914, 915, 916, 917, 918, 919, 920, 921, 922, 923, 924, 925, 926, 927, 928, 929, 930, 931, 932, 933, 934, 935, 936, 937, 938, 939, 940, 941, 942, 943, 944, 945, 946, 947, 948, 949, 950, 951, 952, 953, 954, 955, 956, 957, 958, 959, 960, 961, 962, 963, 964, 965, 966, 967, 968, 969, 970, 971, 972, 974, 975, 976, 977, 978, 979, 980, 981, 982, 983, 984, 985, 986, 987, 988, 989, 990, 991, 992, 993, 994, 995, 996, 997, 998, 999, 1000, 1001, 1002]\n"
     ]
    }
   ],
   "source": [
    "print(list(fin_genotype_df2.index)[:1000] == list(range(1,1000)))\n",
    "print(list(fin_genotype_df2.index)[:1000])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "romance-slovak",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "996\n",
      "{0, 1000, 1002, 1001}\n",
      "{908, 973, 390}\n"
     ]
    }
   ],
   "source": [
    "set1 = set(list(fin_genotype_df2.index)[:1000])\n",
    "set2 = set(range(1,1000))\n",
    "#print(set1)\n",
    "#print(set2)\n",
    "print(len(set1.intersection(set2)))\n",
    "print(set1 - set1.intersection(set2))\n",
    "print(set2 - set1.intersection(set2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "cheap-bridges",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(fin_emission_df2[\"POS\"]) == list(fin_genotype_df2[\"POS\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "unlike-being",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHR</th>\n",
       "      <th>POS</th>\n",
       "      <th>SNP</th>\n",
       "      <th>A2</th>\n",
       "      <th>A1</th>\n",
       "      <th>MAF_AFR</th>\n",
       "      <th>MAF_EUR</th>\n",
       "      <th>AFR_A</th>\n",
       "      <th>AFR_C</th>\n",
       "      <th>AFR_G</th>\n",
       "      <th>AFR_T</th>\n",
       "      <th>EUR_A</th>\n",
       "      <th>EUR_C</th>\n",
       "      <th>EUR_G</th>\n",
       "      <th>EUR_T</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>14</td>\n",
       "      <td>19000017</td>\n",
       "      <td>rs375700886</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>14</td>\n",
       "      <td>19000050</td>\n",
       "      <td>rs543746158</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>14</td>\n",
       "      <td>19000056</td>\n",
       "      <td>rs561973970</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>14</td>\n",
       "      <td>19000059</td>\n",
       "      <td>rs201622908</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.00498</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.995020</td>\n",
       "      <td>0.004980</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>14</td>\n",
       "      <td>19000060</td>\n",
       "      <td>rs28973059</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>0.1188</td>\n",
       "      <td>0.38840</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.881200</td>\n",
       "      <td>0.118800</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.611600</td>\n",
       "      <td>0.388400</td>\n",
       "      <td>0.000001</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   CHR       POS          SNP A2 A1  MAF_AFR  MAF_EUR     AFR_A     AFR_C  \\\n",
       "0   14  19000017  rs375700886  C  T   0.0000  0.00000  0.000001  1.000000   \n",
       "1   14  19000050  rs543746158  G  A   0.0000  0.00000  0.000001  0.000001   \n",
       "2   14  19000056  rs561973970  A  T   0.0000  0.00000  1.000000  0.000001   \n",
       "3   14  19000059  rs201622908  G  T   0.0000  0.00498  0.000001  0.000001   \n",
       "4   14  19000060   rs28973059  C  G   0.1188  0.38840  0.000001  0.881200   \n",
       "\n",
       "      AFR_G     AFR_T     EUR_A     EUR_C     EUR_G     EUR_T  \n",
       "0  0.000001  0.000001  0.000001  1.000000  0.000001  0.000001  \n",
       "1  1.000000  0.000001  0.000001  0.000001  1.000000  0.000001  \n",
       "2  0.000001  0.000001  1.000000  0.000001  0.000001  0.000001  \n",
       "3  1.000000  0.000001  0.000001  0.000001  0.995020  0.004980  \n",
       "4  0.118800  0.000001  0.000001  0.611600  0.388400  0.000001  "
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fin_emission_df2.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "noted-aaron",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "CHR                 14\n",
       "POS           19013506\n",
       "SNP        rs559727658\n",
       "A2                   A\n",
       "A1                   G\n",
       "MAF_AFR       0.000000\n",
       "MAF_EUR       0.000000\n",
       "AFR_A         1.000000\n",
       "AFR_C         0.000001\n",
       "AFR_G         0.000001\n",
       "AFR_T         0.000001\n",
       "EUR_A         1.000000\n",
       "EUR_C         0.000001\n",
       "EUR_G         0.000001\n",
       "EUR_T         0.000001\n",
       "Name: 683, dtype: object"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fin_emission_df2.iloc[682,:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "liable-leader",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "CHR                14\n",
       "POS          19007198\n",
       "SNP        rs28838603\n",
       "A2                  A\n",
       "A1                  T\n",
       "MAF_AFR      0.000000\n",
       "MAF_EUR      0.022910\n",
       "AFR_A        1.000000\n",
       "AFR_C        0.000001\n",
       "AFR_G        0.000001\n",
       "AFR_T        0.000001\n",
       "EUR_A        0.977090\n",
       "EUR_C        0.000001\n",
       "EUR_G        0.000001\n",
       "EUR_T        0.022910\n",
       "Name: 389, dtype: object"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fin_emission_df2.iloc[389,:]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "transsexual-ecuador",
   "metadata": {},
   "source": [
    "Testing pandas.DF.reset_index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "electrical-newman",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2500619\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>CHR</th>\n",
       "      <th>POS</th>\n",
       "      <th>SNP</th>\n",
       "      <th>A2</th>\n",
       "      <th>A1</th>\n",
       "      <th>MAF_AFR</th>\n",
       "      <th>MAF_EUR</th>\n",
       "      <th>AFR_A</th>\n",
       "      <th>AFR_C</th>\n",
       "      <th>AFR_G</th>\n",
       "      <th>AFR_T</th>\n",
       "      <th>EUR_A</th>\n",
       "      <th>EUR_C</th>\n",
       "      <th>EUR_G</th>\n",
       "      <th>EUR_T</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2500614</th>\n",
       "      <td>2505782</td>\n",
       "      <td>14</td>\n",
       "      <td>107289436</td>\n",
       "      <td>rs28971294</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>0.155400</td>\n",
       "      <td>0.3904</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.844600</td>\n",
       "      <td>0.155400</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.609600</td>\n",
       "      <td>0.390400</td>\n",
       "      <td>0.000001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2500615</th>\n",
       "      <td>2505783</td>\n",
       "      <td>14</td>\n",
       "      <td>107289442</td>\n",
       "      <td>rs572272149</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.001786</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.998214</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.001786</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2500616</th>\n",
       "      <td>2505784</td>\n",
       "      <td>14</td>\n",
       "      <td>107289452</td>\n",
       "      <td>rs140896985</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.001786</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.998214</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.001786</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2500617</th>\n",
       "      <td>2505785</td>\n",
       "      <td>14</td>\n",
       "      <td>107289453</td>\n",
       "      <td>rs149770569</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2500618</th>\n",
       "      <td>2505786</td>\n",
       "      <td>14</td>\n",
       "      <td>107289456</td>\n",
       "      <td>rs540661577</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000001</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           index  CHR        POS          SNP A2 A1   MAF_AFR  MAF_EUR  \\\n",
       "2500614  2505782   14  107289436   rs28971294  C  G  0.155400   0.3904   \n",
       "2500615  2505783   14  107289442  rs572272149  A  G  0.001786   0.0000   \n",
       "2500616  2505784   14  107289452  rs140896985  C  T  0.001786   0.0000   \n",
       "2500617  2505785   14  107289453  rs149770569  G  A  0.000000   0.0000   \n",
       "2500618  2505786   14  107289456  rs540661577  C  T  0.000000   0.0000   \n",
       "\n",
       "            AFR_A     AFR_C     AFR_G     AFR_T     EUR_A     EUR_C     EUR_G  \\\n",
       "2500614  0.000001  0.844600  0.155400  0.000001  0.000001  0.609600  0.390400   \n",
       "2500615  0.998214  0.000001  0.001786  0.000001  1.000000  0.000001  0.000001   \n",
       "2500616  0.000001  0.998214  0.000001  0.001786  0.000001  1.000000  0.000001   \n",
       "2500617  0.000001  0.000001  1.000000  0.000001  0.000001  0.000001  1.000000   \n",
       "2500618  0.000001  1.000000  0.000001  0.000001  0.000001  1.000000  0.000001   \n",
       "\n",
       "            EUR_T  \n",
       "2500614  0.000001  \n",
       "2500615  0.000001  \n",
       "2500616  0.000001  \n",
       "2500617  0.000001  \n",
       "2500618  0.000001  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fin_emission_df3, fin_genotype_df3 = standardizeIndices(emission_df_chr14, genotype_df2, \"POS\")\n",
    "print(len(fin_emission_df3))\n",
    "fin_emission_df3.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "hourly-combining",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>CHR</th>\n",
       "      <th>POS</th>\n",
       "      <th>SNP</th>\n",
       "      <th>A2</th>\n",
       "      <th>A1</th>\n",
       "      <th>MAF_AFR</th>\n",
       "      <th>MAF_EUR</th>\n",
       "      <th>AFR_A</th>\n",
       "      <th>AFR_C</th>\n",
       "      <th>AFR_G</th>\n",
       "      <th>AFR_T</th>\n",
       "      <th>EUR_A</th>\n",
       "      <th>EUR_C</th>\n",
       "      <th>EUR_G</th>\n",
       "      <th>EUR_T</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2500614</th>\n",
       "      <td>2505782</td>\n",
       "      <td>14</td>\n",
       "      <td>107289436</td>\n",
       "      <td>rs28971294</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>0.155400</td>\n",
       "      <td>0.3904</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.844600</td>\n",
       "      <td>0.155400</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.609600</td>\n",
       "      <td>0.390400</td>\n",
       "      <td>0.000001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2500615</th>\n",
       "      <td>2505783</td>\n",
       "      <td>14</td>\n",
       "      <td>107289442</td>\n",
       "      <td>rs572272149</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.001786</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.998214</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.001786</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2500616</th>\n",
       "      <td>2505784</td>\n",
       "      <td>14</td>\n",
       "      <td>107289452</td>\n",
       "      <td>rs140896985</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.001786</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.998214</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.001786</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2500617</th>\n",
       "      <td>2505785</td>\n",
       "      <td>14</td>\n",
       "      <td>107289453</td>\n",
       "      <td>rs149770569</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2500618</th>\n",
       "      <td>2505786</td>\n",
       "      <td>14</td>\n",
       "      <td>107289456</td>\n",
       "      <td>rs540661577</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000001</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           index  CHR        POS          SNP A2 A1   MAF_AFR  MAF_EUR  \\\n",
       "2500614  2505782   14  107289436   rs28971294  C  G  0.155400   0.3904   \n",
       "2500615  2505783   14  107289442  rs572272149  A  G  0.001786   0.0000   \n",
       "2500616  2505784   14  107289452  rs140896985  C  T  0.001786   0.0000   \n",
       "2500617  2505785   14  107289453  rs149770569  G  A  0.000000   0.0000   \n",
       "2500618  2505786   14  107289456  rs540661577  C  T  0.000000   0.0000   \n",
       "\n",
       "            AFR_A     AFR_C     AFR_G     AFR_T     EUR_A     EUR_C     EUR_G  \\\n",
       "2500614  0.000001  0.844600  0.155400  0.000001  0.000001  0.609600  0.390400   \n",
       "2500615  0.998214  0.000001  0.001786  0.000001  1.000000  0.000001  0.000001   \n",
       "2500616  0.000001  0.998214  0.000001  0.001786  0.000001  1.000000  0.000001   \n",
       "2500617  0.000001  0.000001  1.000000  0.000001  0.000001  0.000001  1.000000   \n",
       "2500618  0.000001  1.000000  0.000001  0.000001  0.000001  1.000000  0.000001   \n",
       "\n",
       "            EUR_T  \n",
       "2500614  0.000001  \n",
       "2500615  0.000001  \n",
       "2500616  0.000001  \n",
       "2500617  0.000001  \n",
       "2500618  0.000001  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fin_emission_df3.reset_index(inplace=True,drop=True)\n",
    "fin_emission_df3.tail()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "decreased-talent",
   "metadata": {},
   "source": [
    "Testing different indexing strategies too"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "another-candidate",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['index', 'CHR', 'POS', 'SNP', 'A2', 'A1', 'MAF_AFR', 'MAF_EUR', 'AFR_A',\n",
      "       'AFR_C', 'AFR_G', 'AFR_T', 'EUR_A', 'EUR_C', 'EUR_G', 'EUR_T'],\n",
      "      dtype='object')\n",
      "8\n"
     ]
    }
   ],
   "source": [
    "print(fin_emission_df3.columns)\n",
    "em_idx1 = list(fin_emission_df3.columns).index(\"AFR_A\")\n",
    "print(em_idx1)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
